Help command

QueenOfDoom opened the pull request:
Added the help command. I had some thoughts about the command design, since I decided to create sub-commands (based on arguments) in there - and it might be a good idea to extract those into the command template using a `string => function` map.
You'll see when you'll go over it!

QueenOfDoom committed:
Command Template created.

Created a simple Ping Command for Demonstration Purposes

[Bug Fix] Reversed Rate Limit Condition

QueenOfDoom committed:
Refactor: Command Restructuring

Restructured the Command Interface and adapted according files
to the new design!

dkantereivin committed:
refractor,style: autoformat and minor future-proofing

- ran `pnpm prettier:format`
- removed redundant/inferred types
- heavily refractored message.handler.ts

dkantereivin committed:
feat: add permissions infrastructure

dkantereivin committed:
feat: add default command behavior

- includes minor refractor of message.handler.ts

dkantereivin committed:
feat!: switched to es2020 build target

Why the heck didn't we do this in the first place?

dkantereivin committed:
feat!: added ConcreteTrigger objects to .execute()

This commit is a major architectural change to how message handling is handled.
It is UNTESTED.
The idea is to, for all commands, determine the actual triggering content of the message.
This is useful, since functions can now specify why they run, and the
commands can themselves make use of the RegExp matches/function matches and their indices.

- ICommand is now Command, an abstract class
- implemented Command#checkTrigggers(message, keyword, args)
- heavily refractored message.handler.ts
- updated ping.command.ts to avoid breaking changes

dkantereivin committed:
vcs: ignore .idea

Heckin get out thx IntelliJ project config

QueenOfDoom committed:
Merge branch 'master' of https://github.com/ProgrammersPalace/bot

QueenOfDoom committed:
fix: fixed prefix bug

Fixed Prefix Bug, by extracting Prefix Part completely into
the string-trigger condition. Thus args had to be added to
ConcreteTrigger and are now per default the message - split
by spaces. Changes have been made in Ping and Message Handler
to use the changed architecture.

QueenOfDoom committed:
feat: added help command

Added the Help Command and integrated it into the 'Command List'.
It has to be integrated in a special way, due to it registering
every command inside the list on initialization! It has some
sub-commands, which might get extracted in the future. An idea
would be to make it another property `string => function`
inside the command class!

dkantereivin commented:
![image](https://user-images.githubusercontent.com/31638829/112744603-83c80e00-8f6f-11eb-85cc-fecf9701264b.png)
^ first off, you'll want to fix something you're doing on that branch. You're committing lots of extra commits with each PR, so something ain't working properly.

Once you fix that, I'll have another look, and see what actually is being PRed.

QueenOfDoom committed:
feat: added subcommands

Added Subcommands - a new Commands Mechanic, which will enable
you to write `command [sub-command]` and assign a function for
that combination. A default `help`-subcommand is always added,
which prints the syntax of the specified command!
Changes have been made to the Command Constructor, the
`checkTrigger`-function, which now searches recursively for
sub-commands and the changes have been applied to the ping
command and the default behaviour.
I decided to push this change, since sub-commands are a
beautiful way of both writing - as a user and categorizing
commands - as a developer. That way, for example all config
related commands can now be represented using commands like:
- `config get [name]`
- `config set [name] [value]`
- `config list`
- ...

QueenOfDoom committed:
refactor: moved `args` into event-handlers.

Moved `args` from the command/trigger class into the event
handlers!

QueenOfDoom committed:
refactor: applied new command architecture to `help`

Applied the new SubCommand / Syntax Architecture to the
help command!

QueenOfDoom committed:
refactor: applied new command architecture to `help`

Applied the new SubCommand / Syntax Architecture to the
help command!

QueenOfDoom committed:
merge: Upstream -> Local

Just fixed some merge conflicts.

QueenOfDoom committed:
fix: removed merge headers

QueenOfDoom committed:
feat: reaction handler

Added a Reaction Handler, which will react on certain emojis
and redirect them into a check channel. In that check channel
such messages arrive ready to be reviewed and then either
accepted or declined.

QueenOfDoom committed:
Merge branch 'help-command' of https://github.com/QueenOfDoom/bot into help-command

dkantereivin commented:
Closed in favour of corrected #7
